File: core/src/main/java/com/github/jarol/azure/search/spark/sql/connector/core/schema/conversion/input/ReadConverters.java
package com.github.jarol.azure.search.spark.sql.connector.core.schema.conversion.input;

import org.apache.spark.unsafe.types.UTF8String;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.temporal.ChronoUnit;

/**
 * Collection of {@link ReadConverter}(s)
 */

public final class ReadConverters {

    public static final ReadTransformConverter<String> STRING_VALUE_OF;

    public static final ReadTransformConverter<UTF8String> UTF8_STRING;

    /**
     * Converter for int32
     */

    public static final ReadCastConverter<Integer> INT32;

    /**
     * Converter for int64
     */

    public static final ReadCastConverter<Long> INT64;

    /**
     * Converter for doubles
     */

    public static final ReadCastConverter<Double> DOUBLE;

    /**
     * Converter for floats
     */

    public static final ReadCastConverter<Float> SINGLE;

    /**
     * Converter for booleans
     */

    public static final ReadCastConverter<Boolean> BOOLEAN;

    /**
     * Converter for timestamps
     * (internally represented as epoch microseconds)
     */

    public static final ReadTimeConverter<Long> TIMESTAMP;


    /**
     * Converter for days
     * (internally represented as epoch days)
     */

    public static final ReadTimeConverter<Integer> DATE;

    static {

        STRING_VALUE_OF = new ReadTransformConverter<String>() {
            @Override
            protected String transform(Object value) {
                return String.valueOf(value);
            }
        };

        UTF8_STRING = new ReadTransformConverter<UTF8String>() {
            @Override
            protected UTF8String transform(Object value) {
                return UTF8String.fromString((String) value);
            }
        };

        INT32 = new ReadCastConverter<>();
        INT64 = new ReadCastConverter<>();
        DOUBLE = new ReadCastConverter<>();
        SINGLE = new ReadCastConverter<>();
        BOOLEAN = new ReadCastConverter<>();

        TIMESTAMP = new ReadTimeConverter<Long>() {

            @Override
            protected @NotNull Long toInternalObject(
                    @NotNull OffsetDateTime dateTime
            ) {

                return ChronoUnit.MICROS.between(
                        Instant.EPOCH,
                        dateTime.toInstant()
                );
            }
        };

        DATE = new ReadTimeConverter<Integer>() {

            @Override
            protected Integer toInternalObject(
                    OffsetDateTime dateTime
            ) {

                return Long.valueOf(
                        dateTime.toLocalDate().toEpochDay()
                ).intValue();
            }
        };
    }
}

File: core/src/main/scala/com/github/jarol/azure/search/spark/sql/connector/core/schema/conversion/input/ReadConverters.scala
package com.github.jarol.azure.search.spark.sql.connector.core.schema.conversion.input

import com.github.jarol.azure.search.spark.sql.connector.core.Constants
import org.apache.spark.unsafe.types.UTF8String

import java.lang.{Boolean => JBoolean, Double => JDouble, Float => JFloat, Long => JLong}
import java.time.temporal.ChronoUnit
import java.time.{Instant, OffsetDateTime}

/**
 * Converter that applies a transformation from a non-null object to a target type
 * @tparam T target type
 */

trait ReadTransformConverter[T]
  extends ReadConverter {

  override final def apply(value: Any): T = {

    Option(value) match {
      case Some(value) => transform(value)
      case None => null.asInstanceOf[T]
    }
  }

  /**
   * Convert the non-null Search object into a Spark internal object
   * @param value non-null value
   * @return an instance of target type
   */

  protected def transform(value: Any): T
}

/**
 * Converter that applies a simple casting to a non-null value
 * @tparam T target type
 */

final class ReadCastConverter[T]
  extends ReadTransformConverter[T] {
  override protected def transform(value: Any): T = value.asInstanceOf[T]
}

/**
 * Converter for date-time objects
 * @tparam T target type
 */

private sealed trait ReadTimeConverter[T]
  extends ReadTransformConverter[T] {

  override protected final def transform(value: Any): T = {

    // Convert to OffsetDateTime and then transform
    toInternalObject(
      OffsetDateTime.parse(value.asInstanceOf[String],
        Constants.DATETIME_OFFSET_FORMATTER
      )
    )
  }

  /**
   * Convert an instance of [[OffsetDateTime]] to a Spark internal object
   * @param dateTime offset date time
   * @return a Spark internal object
   */

  protected def toInternalObject(dateTime: OffsetDateTime): T
}

object ReadConverters {

  /**
   * Converter for reading numeric values as strings
   */

  final val STRING_VALUE_OF: ReadTransformConverter[String] = (value: Any) => value.asInstanceOf[String]

  /**
   * Converter for strings (internally represented by Spark as [[UTF8String]]s)
   */

  final val UTF8_STRING: ReadTransformConverter[UTF8String] = (value: Any) => UTF8String.fromString(value.asInstanceOf[String])

  final val INT32 = new ReadCastConverter[Integer]

  final val INT64 = new ReadCastConverter[JLong]

  final val DOUBLE = new ReadCastConverter[JDouble]

  final val SINGLE = new ReadCastConverter[JFloat]

  final val BOOLEAN = new ReadCastConverter[JBoolean]

  final val TIMESTAMP: ReadTimeConverter[JLong] = new ReadTimeConverter[JLong] {
    override protected def toInternalObject(dateTime: OffsetDateTime): JLong = {
      ChronoUnit.MICROS.between(Instant.EPOCH, dateTime.toInstant)
    }
  }

  final val DATE: ReadTimeConverter[Integer] = new ReadTimeConverter[Integer] {
    override protected def toInternalObject(dateTime: OffsetDateTime): Integer = {
      dateTime.toLocalDate.toEpochDay.toInt
    }
  }
}
